GH-3846: Document Hazelcast module (#3941)
spring-projects/spring-integration: src/reference/asciidoc/hazelcast.adoc

* `com.hazelcast.topic.MessageListener`

Hazelcast Event-Driven Inbound Channel Adapter listens related cache events and sends event messages to defined channel.
It supports both XML and JavaConfig driven configurations.

==== XML Configuration :

====
[source,xml]
----
<int-hazelcast:inbound-channel-adapter channel="mapChannel"
                      cache="map"
                      cache-events="UPDATED, REMOVED"
                      cache-listening-policy="SINGLE" />
----
====

Basically, Hazelcast Event-Driven Inbound Channel Adapter requires following attributes:

* `channel`: Specifies channel which message is sent;
* `cache`: Specifies the distributed Object reference which is listened.
It is mandatory attribute;
* `cache-events`: Specifies cache events which are listened.
It is optional attribute and its default value is `ADDED`.
Its supported values are as follows :

* Supported cache event types for `IMap` and `MultiMap`: `ADDED`, `REMOVED`, `UPDATED`, `EVICTED`, `EVICT_ALL` and `CLEAR_ALL`;
* Supported cache event types for `ReplicatedMap`: `ADDED`, `REMOVED`, `UPDATED`, `EVICTED`;
* Supported cache event types for `IList`, `ISet` and `IQueue`: `ADDED`, `REMOVED`.

> There is no need to cache event type definition for `ITopic`.

Author: That's correct. There is a dedicated `com.hazelcast.topic.MessageListener` to receive messages published to the topic.

====

It requires four attributes as follows:

* `channel`: Specifies channel which message is sent;
* `cache`: Specifies distributed Map reference which is listened.
Mandatory;
* `cache-events`: Specifies cache events which are listened.
Optional attribute with `ADDED` default value.
Supported values are `ADDED`, `REMOVED`, `UPDATED`, `EVICTED`, `EVICT_ALL` and `CLEAR_ALL`;
* `predicate`: Specifies predicate to listen to the modifications performed on specific map entries.
Mandatory;
* `include-value`: Specifies including of value and oldValue in continuous query result.
Optional with `true` by default;
* `cache-listening-policy`: Specifies cache listening policy as `SINGLE` or `ALL`.
Optional with default value as `SINGLE`.
Each Hazelcast CQ inbound channel adapter listening same cache object with same cache-events attribute, can receive a single event message or all event messages.
If it is `ALL`, all Hazelcast CQ inbound channel adapters listening same cache object with same cache-events attribute, will receive same event messages.
If it is `SINGLE`, they will receive unique event messages.

[[hazelcast-cluster-monitor]]
=== Hazelcast Cluster Monitor Inbound Channel Adapter

> Hazelcast Cluster Monitor enables to listen to the modifications performed on cluster.
> Hazelcast Cluster Monitor Inbound Channel Adapter is an event-driven channel adapter and listens to related Membership, Distributed Object, Migration, Lifecycle and Client events:

Reviewer: ```suggestion
A Hazelcast Cluster Monitor supports listening to modifications performed on the cluster.
The Hazelcast Cluster Monitor Inbound Channel Adapter is an event-driven channel adapter and listens to related Membership, Distributed Object, Migration, Lifecycle and Client events:
```

====
[source, java, role="primary"]
.Java
----
@Bean
public PollableChannel eventChannel() {
    return new QueueChannel();
}

@Bean
public HazelcastInstance hazelcastInstance() {
    return Hazelcast.newHazelcastInstance();
}

@Bean
public HazelcastClusterMonitorMessageProducer hazelcastClusterMonitorMessageProducer() {
    HazelcastClusterMonitorMessageProducer producer = new HazelcastClusterMonitorMessageProducer(hazelcastInstance());
    producer.setOutputChannel(eventChannel());
    producer.setMonitorEventTypes("DISTRIBUTED_OBJECT");

    return producer;
}
----
[source, xml, role="secondary"]
.XML
----
<int:channel id="monitorChannel"/>

<int-hazelcast:cm-inbound-channel-adapter
                 channel="monitorChannel"
                 hazelcast-instance="instance"
                 monitor-types="MEMBERSHIP, DISTRIBUTED_OBJECT" />

<bean id="instance" class="com.hazelcast.core.Hazelcast"
            factory-method="newHazelcastInstance">
    <constructor-arg>
        <bean class="com.hazelcast.config.Config" />
    </constructor-arg>
</bean>
----
====

> It requires four attributes as follows :
> 
> * `channel`: Specifies channel which message is sent;
> * `hazelcast-instance`: Specifies Hazelcast Instance reference to listen cluster events.
> It is mandatory attribute;
> * `monitor-types`: Specifies monitor types which are listened.
> It is optional attribute with `MEMBERSHIP` default value.
> Supported values are `MEMBERSHIP`, `DISTRIBUTED_OBJECT`, `MIGRATION`, `LIFECYCLE`, `CLIENT`.

Reviewer: ```suggestion
It supports three attributes as follows :

* `channel`: Specifies the channel to which messages are sent;
* `hazelcast-instance`: Specifies the Hazelcast Instance reference to listen for cluster events.
It is a mandatory attribute;
* `monitor-types`: Specifies the monitor types which are listened for.
It is an optional attribute with `MEMBERSHIP` being the default value.
Supported values are `MEMBERSHIP`, `DISTRIBUTED_OBJECT`, `MIGRATION`, `LIFECYCLE`, `CLIENT`.
```

[[hazelcast-distributed-sql]]
=== Hazelcast Distributed SQL Inbound Channel Adapter

> Hazelcast allows to run distributed queries on the distributed map.
> Hazelcast Distributed SQL Inbound Channel Adapter is a polling inbound channel adapter.
> It runs defined distributed-sql and returns results in the light of iteration type.

Reviewer: ```suggestion
Hazelcast allows running distributed queries on the distributed map.
The Hazelcast Distributed SQL Inbound Channel Adapter is a polling inbound channel adapter.
It runs the defined distributed-sql command and returns results depending on the iteration type.
```

@Bean
@InboundChannelAdapter(value = "dsDistributedMapChannel", poller = @Poller(maxMessagesPerPoll = "1"))
public HazelcastDistributedSQLMessageSource hazelcastDistributedSQLMessageSource() {
    final HazelcastDistributedSQLMessageSource messageSource =
        new HazelcastDistributedSQLMessageSource(dsDistributedMap(),
            "name='TestName' AND surname='TestSurname'");
    messageSource.setIterationType(DistributedSQLIterationType.ENTRY);

    return messageSource;
}
----
[source, xml, role="secondary"]
.XML
----
<int:channel id="dsMapChannel"/>

<int-hazelcast:ds-inbound-channel-adapter
            channel="dsMapChannel"
            cache="dsMap"
            iteration-type="ENTRY"
            distributed-sql="active=false OR age >= 25 OR name = 'TestName'">
    <int:poller fixed-delay="100"/>
</int-hazelcast:ds-inbound-channel-adapter>

<bean id="dsMap" factory-bean="instance" factory-method="getMap">
    <constructor-arg value="dsMap"/>
</bean>

<bean id="instance" class="com.hazelcast.core.Hazelcast"
            factory-method="newHazelcastInstance">
    <constructor-arg>
        <bean class="com.hazelcast.config.Config" />
    </constructor-arg>
</bean>
----
====

> It requires a poller and four attributes such as:
> 
> * `channel`: Specifies channel which message is sent.
> It is mandatory attribute;
> * `cache`: Specifies distributed `IMap` reference which is queried.
> It is mandatory attribute;
> * `iteration-type`: Specifies result type.
> Distributed SQL can be run on `EntrySet`, `KeySet`, `LocalKeySet` or `Values`.
> It is optional attribute with `VALUE` by default.
> Supported values are `ENTRY, `KEY`, `LOCAL_KEY` and `VALUE`;
> * `distributed-sql`: Specifies where clause of sql statement.
> It is mandatory attribute.

Reviewer: ```suggestion
It requires a poller and supports four attributes:

* `channel`: Specifies the channel to which messages are sent.
It is a mandatory attribute;
* `cache`: Specifies the distributed `IMap` reference which is queried.
It is mandatory attribute;
* `iteration-type`: Specifies result type.
Distributed SQL can be run on `EntrySet`, `KeySet`, `LocalKeySet` or `Values`.
It is an optional attribute with `VALUE` being the default.
Supported values are `ENTRY, `KEY`, `LOCAL_KEY` and `VALUE`;
* `distributed-sql`: Specifies the where clause of the sql statement.
It is a mandatory attribute.
```

Supported Distributed Objects are: `IMap`, `MultiMap`, `ReplicatedMap`, `IList`, `ISet`, `IQueue` and `ITopic`.

====
[source, java, role="primary"]
.Java
----
@Bean
public MessageChannel distributedMapChannel() {
    return new DirectChannel();
}

@Bean
public IMap<Integer, String> distributedMap() {
    return hzInstance().getMap("Distributed_Map");
}

@Bean
public HazelcastInstance hzInstance() {
    return Hazelcast.newHazelcastInstance();
}

@Bean
@ServiceActivator(inputChannel = "distributedMapChannel")
public HazelcastCacheWritingMessageHandler hazelcastCacheWritingMessageHandler() {
    HazelcastCacheWritingMessageHandler handler = new HazelcastCacheWritingMessageHandler();
    handler.setDistributedObject(distributedMap());
    handler.setKeyExpression(new SpelExpressionParser().parseExpression("payload.id"));
    handler.setExtractPayload(true);
    return handler;
}
----
[source, xml, role="secondary"]
.XML
----
<int-hazelcast:outbound-channel-adapter channel="mapChannel"
                    cache-expression="headers['CACHE_HEADER']"
                    key-expression="payload.key"
                    extract-payload="true"/>
----
====

> It requires the following attributes :
> 
> * `channel`: Specifies channel which message is sent;
> * `cache`: Specifies distributed object reference.
> Optional;
> * `cache-expression`: Specifies distributed object via Spring Expression Language(SpEL).
> Optional;
> * `key-expression`: Specifies key of key-value pair via Spring Expression Language(SpEL).
> Optional and required for only for `IMap`, `MultiMap` and `ReplicatedMap` distributed data structures.
> * `extract-payload`: Specifies whole message or just payload to send.
> Optional attribute with `true` by default.
> If it is true, just payload will be written to distributed object.
> Otherwise, whole message will be written by covering both message header and payload.
> 
> By setting distributed object name in the header, messages can be written to different distributed objects via same channel.
> If `cache` or `cache-expression` attributes are not defined, a `HazelcastHeaders.CACHE_NAME` header has to be set in a request `Message`.

Reviewer: ```suggestion
It requires the following attributes :

* `channel`: Specifies the channel to which messages are sent;
* `cache`: Specifies the distributed object reference.
Optional;
* `cache-expression`: Specifies the distributed object via Spring Expression Language (SpEL).
Optional;
* `key-expression`: Specifies the key of a key-value pair via Spring Expression Language (SpEL).
Optional and required for only for `IMap`, `MultiMap` and `ReplicatedMap` distributed data structures.
* `extract-payload`: Specifies whether to send the whole message or just the payload.
Optional attribute with `true` being the default.
If it is true, just the payload will be written to the distributed object.
Otherwise, the whole message will be written by converting both message headers and payload.

By setting distributed object name in the header, messages can be written to different distributed objects via same channel.
If `cache` or `cache-expression` attributes are not defined, a `HazelcastHeaders.CACHE_NAME` header has to be set in a request `Message`.
```

[[hazelcast-leader-election]]
=== Hazelcast Leader Election

If leader election is needed (e.g. for highly available message consumer where only one node should receive messages) a Hazelcast-based `LeaderInitiator` can be used:

====
[source,java]
----
@Bean
public HazelcastInstance hazelcastInstance() {
    return Hazelcast.newHazelcastInstance();
}

@Bean
public LeaderInitiator initiator() {
    return new LeaderInitiator(hazelcastInstance());
}
----
====

> When a node is elected leader it will send `OnGrantedEvent` to all application listeners.

Reviewer: ```suggestion
When a node is elected leader it will send an `OnGrantedEvent` to all application listeners.
```

[[hazelcast-message-store]]
=== Hazelcast Message Store

> For distributed messaging state management, for example for persistent `QueueChannel` or tracking `Aggregator` message groups, the `HazelcastMessageStore` implementation is provided:

Reviewer: ```suggestion
For distributed messaging state management, for example for a persistent `QueueChannel` or tracking `Aggregator` message groups, the `HazelcastMessageStore` implementation is provided:
```

====
[source,java]
----
@Bean
public HazelcastInstance hazelcastInstance() {
    return Hazelcast.newHazelcastInstance();
}

@Bean
public MessageGroupStore messageStore() {
    return new HazelcastMessageStore(hazelcastInstance());
}
----
====

> By default, the `SPRING_INTEGRATION_MESSAGE_STORE` `IMap` is used to store messages and groups key/value manner.

Reviewer: ```suggestion
By default, the `SPRING_INTEGRATION_MESSAGE_STORE` `IMap` is used to store messages and groups as a key/value.
```

Any custom `IMap` can be provided to the `HazelcastMessageStore`.

[[hazelcast-metadata-store]]
=== Hazelcast Metadata Store

An implementation of a `ListenableMetadataStore` is available using a backing Hazelcast `IMap`.

> The default map created with a name `SPRING_INTEGRATION_METADATA_STORE` which can be customized.

Reviewer: ```suggestion
The default map is created with a name `SPRING_INTEGRATION_METADATA_STORE` which can be customized.
```
